### PR TITLE
Backend instrumentation

### PR DESCRIPTION
**Pull request by garimaprachi**

### CODE REVIEW

File: src/AutoWebhook.php [hunk 1]
```diff
@@ -63,6 +64,15 @@ public function autoEnableWebhook($key_id, $key_secret)
                 ['absolute' => TRUE]
             )->toString();
 
+            if($key_id == null || $key_secret == null)
```
Comment (abdulwahidsharief) [inline src/AutoWebhook.php:67]: @garimaprachi use or instead of ||

Comment (abdulwahidsharief) [reply src/AutoWebhook.php:67]: @garimaprachi triggerValidationInstrumentation should be in validateConfigurationForm()

File: drupal_commerce_razorpay.module [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+use Drupal\drupal_commerce_razorpay\Controller\TrackPluginInstrumentation;
+use Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Implements hook_entity_delete().
+ */
+function drupal_commerce_razorpay_entity_delete(EntityInterface $entity) {
+    $query = \Drupal::database()->query("SELECT CAST(`data` AS CHAR(10000) CHARACTER SET utf8) AS decoded_data  FROM config WHERE `name` = :namekey", [':namekey' => 'commerce_payment.commerce_payment_gateway.razorpay']);
```
Comment (abdulwahidsharief) [inline drupal_commerce_razorpay.module:10]: @garimaprachi there are two whitespace before FROM

File: drupal_commerce_razorpay.module [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        
+    $data = unserialize($query->fetchField());
+
+    if(empty($data))
```
Comment (abdulwahidsharief) [inline drupal_commerce_razorpay.module:14]: @garimaprachi leave space after if

File: src/Controller/TrackPluginInstrumentation.php [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            }
+        } catch (\GuzzleHttp\Exception\RequestException $e) {
+            \Drupal::logger('error')->error($e->getMessage());
+        }
+
+        return new Response($response, 200, [
+            'Content-Type' => 'application/json',
+        ]);
+    }
+}
```
Comment (abdulwahidsharief) [inline src/Controller/TrackPluginInstrumentation.php:183]: @garimaprachi leave End of line

File: src/Plugin/Commerce/PaymentGateway/RazorpayCheckout.php [hunk 5]
```diff
@@ -124,26 +125,61 @@ public function validateConfigurationForm(array &$form, FormStateInterface $form
...
+        $query = \Drupal::database()->query("SELECT CAST(`data` AS CHAR(10000) CHARACTER SET utf8) AS decoded_data  FROM config WHERE `name` = :namekey", [':namekey' => 'commerce_payment.commerce_payment_gateway.razorpay']);
+
+        $data = unserialize($query->fetchField());
+
+        $authEvent = '';
+
+        $authProperties = [
+            'is_key_id_populated'     => true,
+            'is_key_secret_populated' => true,
+            'page_url'                => $_SERVER['REQUEST_SCHEME'] . '://' . $_SERVER['HTTP_HOST'].$_SERVER['REQUEST_URI'],
```
Comment (abdulwahidsharief) [inline src/Plugin/Commerce/PaymentGateway/RazorpayCheckout.php:165]: @garimaprachi leave white space before and after concatenation operator

File: drupal_commerce_razorpay.module (hunk 2 above)
Comment (garimaprachi) [reply drupal_commerce_razorpay.module]: Done

File: src/Controller/TrackPluginInstrumentation.php (hunk 4 above)
Comment (garimaprachi) [reply src/Controller/TrackPluginInstrumentation.php]: Done

File: drupal_commerce_razorpay.module (hunk 3 above)
Comment (garimaprachi) [reply drupal_commerce_razorpay.module]: Done

File: src/Controller/TrackPluginInstrumentation.php [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        {
+            \Drupal::logger('error')->error($e->getMessage());
+        }
+    }
+
+    public function getDefaultProperties($timestamp = true)
+    {
+        $defaultProperties = [
+            'platform'            => 'Drupal',
+            'platform_version'    => \Drupal::VERSION,
```
Comment (abdulwahidsharief) [inline src/Controller/TrackPluginInstrumentation.php:146]: @garimaprachi include drupal commerce version

File: src/Controller/TrackPluginInstrumentation.php [hunk 7]
```diff
@@ -0,0 +1,183 @@
+<?php
+
+namespace Drupal\drupal_commerce_razorpay\Controller;
+
+use Razorpay\Api\Api;
+use Razorpay\Api\Errors;
+use Drupal\Core\Http\ClientFactory;
```
[the remark this answers is not in the document]
Comment (garimaprachi) [reply src/Controller/TrackPluginInstrumentation.php]: Done

File: src/AutoWebhook.php [hunk 8]
```diff
@@ -111,23 +121,61 @@ public function autoEnableWebhook($key_id, $key_secret)
...
+                $trackObject = $this->newTrackPluginInstrumentation($key_id, $key_secret);
+                $trackObject->rzpTrackDataLake('autowebhook.created', $webhookProperties);
+
                 \Drupal::logger('RazorpayAutoWebhook')->info('Creating razorpay webhook');
                 return $api->webhook->create($requestBody);
             }
         }
         catch (\Exception $exception)
         {
+            $validationErrorProperties = $this->triggerValidationInstrumentation(
```
[the remark this answers is not in the document]
Comment (garimaprachi) [reply src/AutoWebhook.php]: Removed

File: drupal_commerce_razorpay.install [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+function drupal_commerce_razorpay_install() {
+    $trackObject = new TrackPluginInstrumentation();
+    $trackObject->razorpayPluginActivated();
+}
+
+function drupal_commerce_razorpay_uninstall() {
+    // Clear razorpay data out of the cache.
+    $trackObject = new TrackPluginInstrumentation();
+    $trackObject->razorpayPluginDeactivated();
```
[the remark this answers is not in the document]
Comment (garimaprachi) [reply drupal_commerce_razorpay.install]: Checked is to remain the same for generic purpose

File: src/Controller/TrackPluginInstrumentation.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        $client_factory = \Drupal::service('http_client_factory');
+
+        $client = $client_factory->fromOptions([
+            'headers' => $data['headers'],
+        ]);
+
+        $response = '';
+
+        try {
```
Comment (abdulwahidsharief) [inline src/Controller/TrackPluginInstrumentation.php:151]: @garimaprachi flower bracket should start with new line

File: src/Plugin/Commerce/PaymentGateway/RazorpayCheckout.php [hunk 11]
```diff
@@ -124,26 +125,61 @@ public function validateConfigurationForm(array &$form, FormStateInterface $form
...
+
+        $authProperties = [
+            'is_key_id_populated'     => true,
+            'is_key_secret_populated' => true,
+            'page_url'                => $_SERVER['REQUEST_SCHEME'] . '://' . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'],
+            'auth_successful_status'  => true,
+            'is_plugin_activated'     => $status
+        ];
+
+        if(empty($data['configuration']['key_id']) and empty($data['configuration']['key_secret']))
```
Comment (abdulwahidsharief) [inline src/Plugin/Commerce/PaymentGateway/RazorpayCheckout.php:170]: @garimaprachi leave whitespace after if , every condition should start with new line

**Review by abdulwahidsharief — COMMENTED** [review]
please follow coding standards
https://docs.google.com/document/d/1GZTKZVngt08YzbqvFt15KIHxLXCOo48ly1Q0XSO8rUY/edit#heading=h.7hl6krfv2iid

File: src/Controller/TrackPluginInstrumentation.php (hunk 6 above)
Comment (garimaprachi) [reply src/Controller/TrackPluginInstrumentation.php:129]: Done

**Review by abdulwahidsharief — APPROVED** [review]
LGTM

**Pull request merged**